### ISSUE TITLE
Compare Network Performance

### ISSUE THREAD

**Issue by klannon** [klannon/deep_learning#6]
Create networks using OSU's Torch code and our pylearn2-based code and compare the network performance in terms of classification accuracy.  The final product should be an accuracy vs epoch file for training and testing datasets.

**Comment by mdkdrnevich** [comment]
I'm currently writing an epoch-level callback; what would you expect from this? The idea I had is twofold: 
1. Record the overall current cost of the model on the validation (or test?) data as a two column data file in string format (like what we're reading)
2. Do the equivalent but store the data as a 2D numpy array, which can then be pickled as binary data, which should produce a smaller size data file and ease of retrievability.

Which of these sounds more appealing, or do you have another preference?
Also, should I just be recording the cost vs epoch or do you have another preferred measure of accuracy that should be recorded? Thanks.

**Comment by klannon** [comment]
I'd save the output in a text file for now.  We might want to pass it to other applications (like even Excel, heaven forbid!) and text is more portable.  As for performance measures, OSU uses classification accuracy, which is the fraction of the time where the correct softmax output is the largest one.  E.g. if a training, validation, etc. example is from the class labeled as 1 (of 2) if the first softmax output is the highest, this is counted as a correct classification.  I don't know if pylearn2 already calculates this, but if not, we should do it ourselves for easy comparison with OSU.

**Comment by mdkdrnevich** [comment]
I successfully wrote an algorithm for this performance measure and wanted to give y'all a taste of a test run's output before I finalize my code:
Epochs seen=0
Num Classified = 677801
Epochs seen=1
Num Classified = 907382
Epochs seen=2
Num Classified = 920867
Epochs seen=3
Num Classified = 927327

Looks like good progress.

**Comment by klannon** [comment]
Good progress indeed!  Two questions:
1.  Can you report the numbers separately for the training and testing events?
2.  Can you report fraction classified instead of num?  In other words divide by the size of the sample you considered.

Ideally we'd match this plot from OSU.
![3v_results txt-3](https://cloud.githubusercontent.com/assets/5149849/11129060/5bc94b6a-894b-11e5-8c9e-013898c95a4c.gif)

**Comment by mdkdrnevich** [comment]
Absolutely! I implemented what you have asked and pushed the update to the
repo. Let me know if you need anything else! Are we looking to ROOT for
graphing this data?

**Comment by mdkdrnevich** [comment]
Here is an example of a 35 epoch run on my laptop.

![laptoptest](https://cloud.githubusercontent.com/assets/15619468/11162001/428cbdd2-8a5c-11e5-939a-eedb75b518bc.png)

**Comment by DayOfThePenguin** [comment]
What sort of time per epoch were you getting?  Running on the crc gpu machines (specifically gpu2.crc.nd.edu) has been surprisingly slow (epochs are longer for this dataset, with ~2 million examples than for the SUSY dataset, with 8 million examples).

**Comment by mdkdrnevich** [comment]
I was looking at approximately one minute per epoch, probably an average of
55 seconds. I have attached the log file associated with that graph. I plan
on trying a 55 epoch with my laptop. I have mentioned this to Prof. Lannon
but my laptop has been running at about half the speed of the crc
computers, which implies that either their computers are behind the curve
on processor speed or that cpus run faster than gpus. This is something we
would like to test as well.

**Comment by DayOfThePenguin** [comment]
What are the specs on your laptop?  I'm averaging over 5 minutes per epoch on gpu2 (I am the only job on the machine as well).  I don't think the log file came through

**Comment by DayOfThePenguin** [comment]
I'm getting similar speeds (more than 5 min per epoch) on earth (which is only cpus) as well.  Have you made any changes since your last commit?

**Comment by mdkdrnevich** [comment]
Processor:
Intel(R) Core(TM) i7-5500U CPU @ 2.40GHz, 2394MHz 2 Cores, 4 Logical
Processors
Memory:
8.00 GB
Anything I'm missing?
And no, I haven't made any changes, it should be the same.

**Comment by DayOfThePenguin** [comment]
That's really weird; each of the 12 cores on gpu2 is clocked at 2.4GHz and has 24GB of memory.

**Comment by DayOfThePenguin** [comment]
As I understand it though, openMP is not available on the crc machines by default (it has to be imported 'module load openmp' just like we have to import CUDA), so we are only using one core.

**Comment by klannon** [comment]
@mdkdrnevich:  Any chance you could run your graph out to 100 epochs before 2pm tomorrow and send me the results?

**Comment by DayOfThePenguin** [comment]
Does anyone have any thoughts why we could be getting substantially slower
performance on the CRC machines than on personal computers?

**Comment by mdkdrnevich** [comment]
Here is a 60 epoch test on my laptop, I will try to get a 100 epoch one finished today as well.
![laptoptest](https://cloud.githubusercontent.com/assets/15619468/11253048/5da46f8a-8e07-11e5-8524-8e96a11fad47.png)

**Comment by mdkdrnevich** [comment]
So I was checking my BLAS configuration with Theano and I found their check_blas.py file (theano/misc/check_blas.py) and here is the output.
[myBLASinfo.txt](https://github.com/klannon/deep-learning/files/38430/myBLASinfo.txt)
At least from what I can tell, the BLAS I am using is the same that is linked with numpy....
Any thoughts?

**Comment by mdkdrnevich** [comment]
Here is the 101-epoch run on my laptop. To me it looks like the test data plateaus in accuracy, while the training data begins to overfit to the data. Also, since I completed 100 epochs, and our save setting is every 100 epochs, I also have the model at this point in time pickled, in case you would like to use that for anything.
![laptoptest](https://cloud.githubusercontent.com/assets/15619468/11258530/e156c5e2-8e27-11e5-9d63-e6591a20ad62.png)
